utils/timer/eta.go

```diff
@@ -18,8 +19,109 @@ func ProgressFromHash(b []byte) uint64 {
 	return binary.BigEndian.Uint64(progress[:])
 }
 
-// EstimateETA attempts to estimate the remaining time for a job to finish given
-// the [startTime] and it's current progress.
+// A sample represents a completed amount and the timestamp of the sample
+type sample struct {
+	completed uint64
+	timestamp time.Time
+}
+
+// EtaTracker tracks the ETA of a job
+type EtaTracker struct {
+	samples        []sample
+	samplePosition uint8
+	maxSamples     uint8
+	totalSamples   uint64
+	slowdownFactor float64
+}
+
+// NewEtaTracker creates a new EtaTracker with the given maximum number of samples
+// and a slowdown factor. The slowdown factor is a multiplier that is added to the ETA
+// based on the percentage completed.
+//
+// The adjustment works as follows:
+//   - At 0% progress: ETA is multiplied by slowdownFactor
+//   - At 100% progress: ETA is the raw estimate (no adjustment)
+//   - Between 0% and 100%: Adjustment decreases linearly with progress
+//
+// Example: With slowdownFactor = 2.0:
+//   - At 0% progress: ETA = raw_estimate * 2.0
+//   - At 50% progress: ETA = raw_estimate * 1.5
+//   - At 100% progress: ETA = raw_estimate * 1.0
+//
+// If maxSamples is less than 1, it will default to 5
+func NewEtaTracker(maxSamples uint8, slowdownFactor float64) *EtaTracker {
```

Comment:
Are slowdown factors normal for ETAs? It feels really odd to me that we aren't just calculating the current rate of progress and extrapolating.